lap3dkernel/c++SIMD/lap3dkernel.cpp

```diff
@@ -126,6 +166,45 @@ void test_kernel_vec256(const Real* Xs, const Real* Ys, const Real* Zs,
         pot_vec.store(pot + t);
     }
 }
+
+void test_kernel_vec256_fast_rsqrt(const Real* Xs, const Real* Ys, const Real* Zs,
+                    int ns, const Real* charge,  
+                 const Real* Xt, const Real* Yt, const Real* Zt, int nt, Real* pot){
+#pragma omp parallel for
+    for(int t = 0; t < nt; t += 4){
+
+      Vec4d Xt_vec, Yt_vec, Zt_vec, pot_vec;
+        // load
+        Xt_vec.load(Xt + t);
+        Yt_vec.load(Yt + t);
+        Zt_vec.load(Zt + t);
+        pot_vec.load(pot + t);
+        
+        for(int s = 0; s < ns; s++){
+            Vec4d dx = Vec4d(Xs[s]) - Xt_vec;
+            Vec4d dy = Vec4d(Ys[s]) - Yt_vec;
+            Vec4d dz = Vec4d(Zs[s]) - Zt_vec;
+            Vec4d r2 = dx*dx+dy*dy+dz*dz;
+            
+            // Newton iterations for rsqrt double precision
+            Vec4d rinv = approx_rsqrt(r2);
+            // Note the trick here changes the Newton iteration constants
+            // the standard two Newton iterations are,
+            // rinv *= ((3.0) - r2 * rinv * rinv) * 0.5;
+            // rinv *= ((3.0) - r2 * rinv * rinv) * 0.5;
+            // we can save one simd multiplication " * 0.5 " by using the following trick
+            // first Newton iteration
+            rinv *= ((3.0) - r2 * rinv * rinv);
+            // second Newton iteration
+            rinv *= ((12.0) - r2 * rinv * rinv) * (0.0625);
```

Comment:
The modified Newton iteration saves one simd multiplication, the standard Newton iteration should be
rinv *= ((3.0) - r2 * rinv * rinv) * 0.5